pkg/controller/jobs/statefulset/statefulset_reconciler.go

```diff
@@ -144,15 +159,53 @@ func (r *Reconciler) Update(e event.UpdateEvent) bool {
 	return r.handle(e.ObjectNew)
 }
 
-func (r *Reconciler) Delete(event.DeleteEvent) bool {
-	return false
+func (r *Reconciler) Delete(e event.DeleteEvent) bool {
+	return r.handle(e.Object)
 }
 
 func (r *Reconciler) handle(obj client.Object) bool {
 	sts, isSts := obj.(*appsv1.StatefulSet)
 	if !isSts {
-		return false
+		return true
 	}
 	// Handle only statefulset managed by kueue.
 	return jobframework.IsManagedByKueue(sts)
 }
+
+var _ handler.EventHandler = (*podHandler)(nil)
+
+type podHandler struct{}
+
+func (h *podHandler) Generic(context.Context, event.GenericEvent, workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+}
+
+func (h *podHandler) Create(_ context.Context, e event.CreateEvent, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	// To watch for pod ADDED events.
+	h.handle(e.Object, q)
+}
+
+func (h *podHandler) Update(_ context.Context, e event.UpdateEvent, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	h.handle(e.ObjectNew, q)
+}
+
+func (h *podHandler) Delete(context.Context, event.DeleteEvent, workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+}
+
+func (h *podHandler) handle(obj client.Object, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod || pod.Annotations[podcontroller.SuspendedByParentAnnotation] != FrameworkName {
+		return
+	}
+	if controllerRef := metav1.GetControllerOf(pod); controllerRef != nil {
+		if controllerRef.Kind != gvk.Kind || controllerRef.APIVersion != gvk.GroupVersion().String() {
+			// The pod is controlled by an owner that is not an apps/v1 StatefulSet.
```

Comment:
The owner of the Pod is never StatefulSet, so I'm confused by the comment actually.

I think [here](https://github.com/kubernetes-sigs/kueue/blob/577474373fafcd953faae122c7797a5ca2a7dad4/pkg/controller/jobs/pod/pod_webhook.go#L173-L186) is the analogous problem solved properly. It would be nice to commonize the code.

Reply:
I created a StatefulSet without Kueue and observed the same behavior:

```
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: "2025-02-10T11:56:26Z"
  generateName: nginx-statefulset-
  labels:
    app: nginx
    apps.kubernetes.io/pod-index: "0"
    controller-revision-hash: nginx-statefulset-89458d665
    statefulset.kubernetes.io/pod-name: nginx-statefulset-0
  name: nginx-statefulset-0
  namespace: default
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: StatefulSet
    name: nginx-statefulset
    uid: 37acd725-d7ea-4039-8ad0-1c78d375e65c
  resourceVersion: "513"
  uid: a0d1f759-be3b-4b2b-a96f-894018a67125
```

We are not setting ownerReference on the Pod.

Reply:
I see, indeed, it seems ReplicaSets are only used for Deployements and not for StatefulSets, so all good.

Reply:
Yes, the relationship for each resources are

- Deployment -> ReplicaSet -> Pod
- StatefulSet -> Pod
- DaemonSet -> Pod
- Job -> Pod